### PR TITLE
Fixed .NET SDK style project slowness in the editor

### PR DESCRIPTION
**Pull request by TIHan**
This was something; took quite a bit of time tracking down, but now it's starting to make sense.

**Edited**

We changed the approach on how we fix this problem. Essentially, we are removing calls to `UpdateProjectInfo` on document opened/changed.

Test Plan

- New .NET SDK style project, measure time to first semantic highlighting
- Edit document in .NET SDK style project, time to semantic highlighting and autocomplete
- Generate large .NET SDK style solutions under `tests/projects/stress` and check load time and first semantic highlighting
- Check script in .NET SDK style project with `Include=None`.
- Check script in legacy project with `Include=None`
- Check script with same cases as above, but with `Include=Compile`
- Check script in no project without any project loaded
- Check out of project `.fs` file without a project loaded
- Check out of project `.fs` file where a project loaded .net sdk / legacy
- Check out of project `.fsx` script file where a project loaded .net sdk / legacy
- Close and re-open `SortedMap.fs` in Spreads, time to semantic highlighting after re-open.
- Close and re-open huge generated .NET SDK style `.fs` file, time to semantic highlighting after re-open.
- Add file in .NET SDK style project, check semantic highlighting
- Remove file in .NET SDK style project, check semantic highlighting

### CODE REVIEW

**Comment by wallymathieu** [comment]
Nice 👍

**Comment by vasily-kirichenko** [comment]
This code deserves a small unit test :)

**Comment by forki** [comment]
But first merge it so that it's going into 15.7.2 ;-)

**Comment by dsyme** [comment]
Great work.

**Comment by dsyme** [comment]
Could you take a while to think how to put this under automated test? It's important never to regress this... 

The test suite has oodles of tests to check we dont reconpute/invalidate, counting number of parses, checks etc. But that only works if you get the project options right.

**Comment by TIHan** [comment]
I agree we should put a test here so we don't regress. At the top of my head, I'm not sure how to accomplish this as it's really in the middle of a lot of stuff. Unless we already have mocks, this will be quite a bit of work to do. 

At some point we are going to refactor parts of the integration with the workspace. Perhaps during that time we can design something that is testable, isolated from the language service and workspace.

**Comment by TIHan** [comment]
I'll try to think of something.

**Comment by TIHan** [comment]
As for `UpdateProjectInfo`, @KevinRansom will have more context on why it's done the way it's done.

**Comment by dsyme** [comment]
> As for UpdateProjectInfo, @KevinRansom will have more context on why it's done the way it's done.

I wrote the code originally, so I know the context.  Looking at the code now.

Do you have a callstack for when UpdateProjectInfo is being called during standard editing?

**Comment by TIHan** [comment]
I don't on hand. I can get one next time I go into the office.

**Comment by TIHan** [comment]
Your best bet is to look at the workspace events and see how it eventually calls UpdateProjectInfo.

**Comment by dsyme** [comment]
Oh I see, yes.  The following path from `OnDocumentChanged` to `UpdateProjectInfo` is just plain wrong - we should never have been updating the project info on every change to the document.  

May I mention that I [asked this question when reviewing this code](https://github.com/Microsoft/visualfsharp/pull/4121#issuecomment-352763591), when @cartermp proposed a fix for another manifestation of this problem in https://github.com/Microsoft/visualfsharp/pull/4121:

```
    member private this.OnDocumentChanged(projectId:ProjectId, documentId:DocumentId) = projectInfoManager.UpdateDocumentInfoWithProjectId(projectId, documentId, "OnDocumentChanged", invalidateConfig=false)

    member this.UpdateDocumentInfoWithProjectId(projectId:ProjectId, documentId:DocumentId, userOpName, invalidateConfig) =
        if workspace.IsDocumentOpen(documentId) then
            this.UpdateProjectInfoWithProjectId(projectId, userOpName, invalidateConfig)

    member this.UpdateProjectInfoWithProjectId(projectId:ProjectId, userOpName, invalidateConfig) =
        let hier = workspace.GetHierarchy(projectId)
        match hier with
        | null -> ()
        | h when (h.IsCapabilityMatch("CPS")) ->
            let project = workspace.CurrentSolution.GetProject(projectId)
            if not (isNull project) then
                let siteProvider = this.ProvideProjectSiteProvider(project)
                let projectSite = siteProvider.GetProjectSite()
                if projectSite.CompilationSourceFiles.Length <> 0 then
                    this.UpdateProjectInfo(tryGetOrCreateProjectId, projectId, projectSite, userOpName, invalidateConfig)
        | _ -> ()
```

> @KevinRansom Do you know why we are doing anything at all for DocumentChanged? Was there a particular case that motivated this? e.g. why do we need to call UpdateDocumentInfoWithProjectId at all for OnDocumentChanged?

However I got no reply - @KevinRansom please make sure to keep track of questions like this and make sure they get a reply?

Anyway, I strongly believe that we that we should simply **not** be updating the project info on DocumentChanged.  The only case where we might do that is for scripts, but they are handled differently.

**Comment by dsyme** [comment]
Anyway, I believe the correct fix is just to delete the event handler for OnDocumentChanged.  And if that isn't a correct fix (I think it is) then we really need to work out why the event handler is needed....

**Comment by TIHan** [comment]
Deleting it would be ideal. We should also delete it from the `textView.GetBuffer` as well of we can. Opening a document shouldn't so that either, legacy doesn't.

**Comment by dsyme** [comment]
@vasily-kirichenko 

> Can it result with loss of updating project options for FSX files when #r directives added/deleted?

Good question. Script files are covered by this code in `TryGetOptionsForDocumentOrProject`:
```
            // The options for a single-file script project are re-requested each time the file is analyzed.  This is because the
            // single-file project may contain #load and #r references which are changing as the user edits, and we may need to re-analyze
            // to determine the latest settings.  FCS keeps a cache to help ensure these are up-to-date.
            match singleFileProjectTable.TryGetValue(projectId) with
            | true, (loadTime, _, _) ->
                ...
```
But the behaviour you mention should be checked as part of validating the fix .

> We should also delete it from the textView.GetBuffer as well if we can.

@TIHan do you have a link for that?

**Comment by TIHan** [comment]
@vasily-kirichenko that's an interesting point. I wonder if adding `#r`'s with the current fix will work or not.

For script files, we should be using the [MiscellaneousFilesWorkspace](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs,344ad90895765109) which we are currently not. If we used that workspace, we could at least separate its specific behavior from the workspace for projects.

**Comment by TIHan** [comment]
It's not super clear but:
This is the legacy path: https://github.com/TIHan/visualfsharp/blob/a6fc8525b8fec6affdef69361d3847829b9d4314/vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs#L669

And this is the CPS path:
https://github.com/TIHan/visualfsharp/blob/a6fc8525b8fec6affdef69361d3847829b9d4314/vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs#L671

**Comment by dsyme** [comment]
> For script files, we should be using the MiscellaneousFilesWorkspace which we are currently not. If we used that workspace, we could at least separate its specific behavior from the workspace for projects.

Yeah, there are a couple of glitches with script files (e.g. loss of colorization on save-as file rename) that can likely be fixed by updating script file options on changes like these, or by doing things differently in the first place
```
            | WorkspaceChangeKind.AdditionalDocumentAdded
            | WorkspaceChangeKind.AdditionalDocumentReloaded
            | WorkspaceChangeKind.AdditionalDocumentRemoved
```

Re SetupNewTextView - perhaps that is just the wrong place to be setting up the options for our-of-project files and scripts, and that doing it in response to the events above will be more complete and accurate (and more portable too - we'd like to eventually use FSharp.Editor in VS for Mac, so the less VS-specific code the better. Roslyn and Roslyn CPS-specific code is ok).

**Comment by TIHan** [comment]
> Yeah, there are a couple of glitches with script files (e.g. loss of colorization on save-as file rename) that can likely be fixed by updating script file options on changes like these, or by doing things differently in the first place

Then this fix needs to evolve a bit more. Sorry I didn't test that case.
Does the `AdditionalDocument` refer to files that are not part of the project, such as our script files maybe?

> doing it in response to the events above will be more complete and accurate

This is exactly what we need to do. We should always listen to the workspace to give us what we need. Right now we rely on HandleCommandLineArgs method until we figure out how to get rid of it. We need to integrate with the evaluation/design-time builds.

**Comment by dsyme** [comment]
> We should also delete it from the textView.GetBuffer as well if we can.

OK, I understand now. I added some comments to clarify the existing code [here](https://github.com/Microsoft/visualfsharp/pull/4891/files).  Could you check the comments, particularly [this one](https://github.com/Microsoft/visualfsharp/pull/4891/files#diff-cdf0a4fbb7615b8029d78133924259c3R691)?

Let me explain a bit about `useUniqueStamp`. This stems from the time when ProjectSitesAndFiles.fs was shared between `FSharp.Editor` and `FSharp.LanguageService`.  For `FSharp.Editor` this is always *true*, (and for the legacy `FSharp.LanguageService` this was always *false*).  So the meaning of the flag is "are we using the unique stamping facility at all".  Since we no longer share this file we can remove this flag, assuming it always to be true.

**Comment by dsyme** [comment]
> > Yeah, there are a couple of glitches with script files (e.g. loss of colorization on save-as file rename) that can likely be fixed by updating script file options on changes like these, or by doing things differently in the first place

> Then this fix needs to evolve a bit more. Sorry I didn't test that case.

Oh, they are existing glitches, not glitches in this PR.  See https://github.com/Microsoft/visualfsharp/issues/1944

> Does the AdditionalDocument refer to files that are not part of the project, such as our script files maybe?

I believe so, but should be checked.

**Comment by dsyme** [comment]
@TIHan I removed the `useUniqueStamp` flag as part of https://github.com/Microsoft/visualfsharp/pull/4891, to clarify.

**Comment by dsyme** [comment]
@saul Am I correct in thinking that `WorkspaceChangeKind.AdditionalDocumentAdded/Removed` is the correct thing, rather than watching `MiscellaneousFilesWorkspace` per se?  That is

* scripts that are part of a project and not `Compiled` will get "AdditionalDocument" events in a normal workspace
* scripts that are part of a project and marked `Compiled` be get normal "Document" events in a normal workspace
* scripts that are not part of a project will be "AdditionalDocument" in "MiscellaneousFilesWorkspace"

@TIHan FWIW I don't think any changes to do with scripts need be part of this PR, which should focus on the problem with DocumentChanged event within projects first

**Comment by saul** [comment]
@dsyme I’m not too sure. Unfortunately I haven’t looked at this in about a year.

**Comment by TIHan** [comment]
PR is updated. I'm going to add a generated netcore solution so we can easily test this.

**Comment by TIHan** [comment]
Tests have been added.

**Comment by TIHan** [comment]
test Windows_NT Release_ci_part2 Build please

**Comment by forki** [comment]
So the solution was to delete code that was already flag with weird comment?

**Comment by forki** [comment]
@cartermp it's too late for 15.7.2?

**Comment by cartermp** [comment]
@forki It's highly likely that it's too late for 15.7.2. Approvals for bugs we wanted to take for that went out last week.

**Comment by cartermp** [comment]
Another valueable one to user test is SQLProvider: https://github.com/fsprojects/SQLProvider

It's definitely suffering today

**Comment by dsyme** [comment]
@cartermp Are nightlies now usable with dev 15.7?  We should put out some kind of announcement when that switch happens? (I'm not on dev 15.7 yet)

**Comment by cartermp** [comment]
Yes for both: https://dotnet.myget.org/feed/fsharp/package/vsix/VisualFSharp

We're producing 15.7.x versions for master and a build of a VSIX ought to do the same.

**Comment by TIHan** [comment]
Doing some more testing today. So the `MruCache` for incremental builder is still getting invalidated, but this time after we get completions/intellisense. Something else is causing this, I'm going to investigate this.

**Comment by TIHan** [comment]
Ok, I think I know what's happening from my last comment.

It's because of multi-targetting. Mutli-targetted projects, are treated as two separate logical projects in the workspace. My hypothesis is that FCS only keys off of the project file name, rather than project file name and TFM. When this happens, there could be a race condition, and depending on the race condition will kick off invalidation and rebuild the world.

So this is a bug already today and is not from our current changes made here. Perhaps I can make another PR to fix multi-targeting.

**Comment by Pilchie** [comment]
Just a note about `AdditionalFiles` - they are *not* files that are outside of projects like scripts.  Instead they are "extra" files added to to the compilation to be used as input to analyzers.  Things like the "PublicAPI.unshipped.txt" in the roslyn repo.

As @saul said, we should *definitely* be hooking into `MiscellaneousFilesWorkspace` instead of doing this in `SetupNewView`.

**Comment by Pilchie** [comment]
Couple of questions:

1. If the test code is just generated boilerplate, why not generate it on each run instead of checking it in?
2. What's with the C# only solutions? 
3. Can you point me to where this stuff is actually exercised, or is that still a manual process?

File: vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs [hunk 1]
```diff
@@ -439,7 +438,6 @@ type internal FSharpLanguageService(package : FSharpPackage) =
             | WorkspaceChangeKind.ProjectAdded     -> this.OnProjectAdded(args.ProjectId)
             | WorkspaceChangeKind.ProjectReloaded  -> this.OnProjectReloaded(args.ProjectId)
             | WorkspaceChangeKind.DocumentAdded    -> this.OnDocumentAdded(args.ProjectId, args.DocumentId)
-            | WorkspaceChangeKind.DocumentChanged  -> this.OnDocumentChanged(args.ProjectId, args.DocumentId)
```
Comment (Pilchie) [inline vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs]: Strange that `DocumentChanged` was in this list twice already.  Also - consider whether `DocumentInfoChanged` should result in a call.

**Comment by forki** [comment]
The advantage of checking it in is that contributors usually have it easier
to use it. Also we have history of it (which can also be seen as a downside
;-))

Will Smith <notifications@github.com> schrieb am Mo., 14. Mai 2018, 18:24:

**Comment by TIHan** [comment]
I'm removing the generated projects. I think it is fine to run the generation manually by invoking fsi. This keeps the commits much clearer than seeing 100s of generated files being committed. If we really want the generated projects to be committed, we can make a separate PR.

File: vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs (hunk 1 above)
Comment (TIHan) [reply vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs]: Thinking about this, once we move away from handlecommandlinargs, we will need to listen to DocumentInfoChanged most likely.

**Comment by dsyme** [comment]
> I'm removing the generated projects. I think it is fine to run the generation manually by invoking fsi. This keeps the commits much clearer than seeing 100s of generated files being committed. If we really want the generated projects to be committed, we can make a separate PR.

Yes make a separate PR please. The huge advantage of having them in the tree is that everyone, everywhere has **exactly** the same baselines to test against for VS perf, without any effort. That's incredibly valuable.

File: vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs (hunk 1 above)
Comment (dsyme) [reply vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs]: We didn't see any DocumentInfoChanged  events - when do they arise?

**Comment by forki** [comment]
As said above +1 for checking them in. We already saw benefits of checked in test projects in older issues.

**Comment by vasily-kirichenko** [comment]
I’m for checking them in too. Another reason: the generation script does not work on Mac out of the box due to the windows style slashes in paths.

File: vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs (hunk 1 above)
Comment (Pilchie) [reply vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs]: For example when the filename changes, etc:
```C#
        /// <summary>
        /// The document in the current solution had is info changed; name, folders, filepath
        /// </summary>
        DocumentInfoChanged = 17,
```

**Pull request merged**